File: 11.containerWithMostWater/containerWithMostWater.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution{
public:
  int maxArea(vector<int>& height) {
    int left = 0, right = height.size() - 1;
    int maxArea = 0;
    while(left < right) {
      maxArea = max(maxArea,  (right - left) * min(height[left], height[right]));
      if(height[left] > height[right])
	 right--;
      else
	left++;	   
    }
    return maxArea;
  }
};

int main()
{
  Solution s;
  int a[4] = {1,2,3,1};
  vector<int> vec(a, a + 4);
  cout << s.maxArea(vec) << endl;
  return 0;
}

File: 11.containerWithMostWater/containerWithMostWater.java
class Solution {
    public int maxArea(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int res = 0;
        while (left < right) {
            int width = right - left;
            if (height[left] < height[right]) {
                res = Math.max(res, (height[left] * width));
                left++;
            } else {
                res = Math.max(res, (height[right] * width));
                right--;
            }
        }
        return res;
    }
}